File: sdk/typescript/src/a2a/resource.ts
/** Inkbox serve-side A2A inbox resource. */

import type { HttpTransport } from "../_http.js";
import type { FilterMode } from "../mail/types.js";
import type {
  A2AContactRule,
  A2AContext,
  A2AContextPage,
  A2AHistoryMessage,
  A2AHistoryMessagePage,
  A2AMessageListOptions,
  A2AReplyIntent,
  A2ARuleAction,
  A2ARuleDirection,
  A2ASettings,
  A2ASkill,
  A2ATask,
  A2ATaskListOptions,
  A2ATaskPage,
  A2ASentTaskListOptions,
} from "./types.js";
import {
  parseA2AContext,
  parseA2AHistoryMessage,
  parseA2ATask,
} from "./types.js";

type Raw = Record<string, any>;

function base(handle: string): string {
  return `/identities/${encodeURIComponent(handle)}/a2a`;
}

function ruleDirection(direction: A2ARuleDirection): "inbound" | "both" {
  if (direction !== "inbound" && direction !== "both") {
    throw new TypeError("A2A rule direction must be 'inbound' or 'both'");
  }
  return direction === "inbound" ? "inbound" : "both";
}

function parseSettings(raw: Raw): A2ASettings {
  return {
    enabled: raw.enabled,
    filterMode: raw.filter_mode,
    skills: raw.skills ?? null,
    cardUrl: raw.card_url,
    inboundTaskCount: raw.inbound_task_count ?? 0,
    outboundTaskCount: raw.outbound_task_count ?? 0,
    updatedAt: raw.updated_at ?? null,
  };
}

function parseRule(raw: Raw): A2AContactRule {
  return {
    id: raw.id,
    action: raw.action,
    matchType: raw.match_type,
    matchTarget: raw.match_target,
    direction: raw.direction,
    status: raw.status,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

export class A2AResource {
  constructor(private readonly http: HttpTransport) {}

  async settings(handle: string): Promise<A2ASettings> {
    return parseSettings(await this.http.get<Raw>(`${base(handle)}/settings`));
  }

  async updateSettings(
    handle: string,
    changes: { enabled?: boolean; filter_mode?: FilterMode; skills?: A2ASkill[] | null },
  ): Promise<A2ASettings> {
    return parseSettings(
      await this.http.put<Raw>(`${base(handle)}/settings`, changes),
    );
  }

  card(handle: string): Promise<Record<string, unknown>> {
    return this.http.get(`${base(handle)}/card`);
  }

  async tasks(
    handle: string,
    options: A2ATaskListOptions = {},
  ): Promise<A2ATaskPage> {
    const raw = await this.http.get<Raw>(`${base(handle)}/tasks`, {
      direction: options.direction,
      requester_handle: options.requesterHandle,
      worker_handle: options.workerHandle,
      state: options.state,
      context_id: options.contextId,
      q: options.q,
      since: options.since,
      cursor: options.cursor,
      limit: options.limit ?? 50,
    });
    return {
      items: (raw.items ?? []).map(parseA2ATask),
      nextCursor: raw.next_cursor ?? null,
    };
  }

  async *iterTasks(
    handle: string,
    options: Omit<A2ATaskListOptions, "cursor"> = {},
  ): AsyncGenerator<A2ATask> {
    let cursor: string | undefined;
    do {
      const page = await this.tasks(handle, { ...options, cursor });
      yield* page.items;
      cursor = page.nextCursor ?? undefined;
    } while (cursor);
  }

  async task(handle: string, taskId: string): Promise<A2ATask> {
    return parseA2ATask(
      await this.http.get<Raw>(`${base(handle)}/tasks/${taskId}`),
    );
  }

  async sentTasks(
    handle: string,
    options: A2ASentTaskListOptions = {},
  ): Promise<A2ATaskPage> {
    const raw = await this.http.get<Raw>(`${base(handle)}/sent/tasks`, {
      requester_handle: options.requesterHandle,
      worker_handle: options.workerHandle,
      state: options.state,
      context_id: options.contextId,
      q: options.q,
      since: options.since,
      cursor: options.cursor,
      limit: options.limit ?? 50,
    });
    return {
      items: (raw.items ?? []).map(parseA2ATask),
      nextCursor: raw.next_cursor ?? null,
    };
  }

  async *iterSentTasks(
    handle: string,
    options: Omit<A2ASentTaskListOptions, "cursor"> = {},
  ): AsyncGenerator<A2ATask> {
    let cursor: string | undefined;
    do {
      const page = await this.sentTasks(handle, { ...options, cursor });
      yield* page.items;
      cursor = page.nextCursor ?? undefined;
    } while (cursor);
  }

  async sentTask(handle: string, taskId: string): Promise<A2ATask> {
    return parseA2ATask(
      await this.http.get<Raw>(`${base(handle)}/sent/tasks/${taskId}`),
    );
  }

  async messages(
    handle: string,
    options: A2AMessageListOptions = {},
  ): Promise<A2AHistoryMessagePage> {
    const raw = await this.http.get<Raw>(`${base(handle)}/messages`, {
      direction: options.direction,
      requester_handle: options.requesterHandle,
      worker_handle: options.workerHandle,
      task_id: options.taskId,
      context_id: options.contextId,
      role: options.role,
      q: options.q,
      since: options.since,
      cursor: options.cursor,
      limit: options.limit ?? 50,
    });
    return {
      items: (raw.items ?? []).map(parseA2AHistoryMessage),
      nextCursor: raw.next_cursor ?? null,
    };
  }

  async *iterMessages(
    handle: string,
    options: Omit<A2AMessageListOptions, "cursor"> = {},
  ): AsyncGenerator<A2AHistoryMessage> {
    let cursor: string | undefined;
    do {
      const page = await this.messages(handle, { ...options, cursor });
      yield* page.items;
      cursor = page.nextCursor ?? undefined;
    } while (cursor);
  }

  async reply(
    handle: string,
    taskId: string,
    options: {
      intent: A2AReplyIntent;
      text?: string;
      parts?: Record<string, unknown>[];
    },
  ): Promise<A2ATask> {
    if ((options.text === undefined) === (options.parts === undefined)) {
      throw new TypeError("Pass exactly one of text or parts");
    }
    return parseA2ATask(
      await this.http.post<Raw>(`${base(handle)}/tasks/${taskId}/reply`, {
        intent: options.intent,
        parts: options.text === undefined ? options.parts : [{ text: options.text }],
      }),
    );
  }

  async contexts(
    handle: string,
    options: {
      direction?: "inbound" | "outbound" | "both";
      cursor?: string;
      limit?: number;
    } = {},
  ): Promise<A2AContextPage> {
    const raw = await this.http.get<Raw>(`${base(handle)}/contexts`, {
      direction: options.direction,
      cursor: options.cursor,
      limit: options.limit ?? 50,
    });
    return {
      items: (raw.items ?? []).map(parseA2AContext),
      nextCursor: raw.next_cursor ?? null,
    };
  }

  async context(handle: string, contextId: string): Promise<A2AContext> {
    return parseA2AContext(
      await this.http.get<Raw>(`${base(handle)}/contexts/${contextId}`),
    );
  }

  async sentContexts(
    handle: string,
    options: { cursor?: string; limit?: number } = {},
  ): Promise<A2AContextPage> {
    const raw = await this.http.get<Raw>(`${base(handle)}/sent/contexts`, {
      cursor: options.cursor,
      limit: options.limit ?? 50,
    });
    return {
      items: (raw.items ?? []).map(parseA2AContext),
      nextCursor: raw.next_cursor ?? null,
    };
  }

  async sentContext(handle: string, contextId: string): Promise<A2AContext> {
    return parseA2AContext(
      await this.http.get<Raw>(
        `${base(handle)}/sent/contexts/${contextId}`,
      ),
    );
  }

  async contactRules(handle: string): Promise<A2AContactRule[]> {
    return (
      await this.http.get<Raw[]>(`${base(handle)}/contact-rules`)
    ).map(parseRule);
  }

  async addContactRule(
    handle: string,
    options: {
      handle: string;
      action: A2ARuleAction;
      direction?: A2ARuleDirection;
    },
  ): Promise<A2AContactRule> {
    return parseRule(
      await this.http.post<Raw>(`${base(handle)}/contact-rules`, {
        action: options.action,
        match_type: "handle",
        match_target: options.handle,
        direction: ruleDirection(options.direction ?? "inbound"),
      }),
    );
  }

  async updateContactRule(
    handle: string,
    ruleId: string,
    options: {
      action?: A2ARuleAction;
      direction?: A2ARuleDirection;
    },
  ): Promise<A2AContactRule> {
    if (options.action === undefined && options.direction === undefined) {
      throw new TypeError("Pass at least one of action or direction");
    }
    const body = {
      action: options.action,
      direction: options.direction === undefined
        ? undefined
        : ruleDirection(options.direction),
    };
    return parseRule(
      await this.http.patch<Raw>(
        `${base(handle)}/contact-rules/${ruleId}`,
        body,
      ),
    );
  }

  async deleteContactRule(handle: string, ruleId: string): Promise<void> {
    await this.http.delete(`${base(handle)}/contact-rules/${ruleId}`);
  }
}

File: sdk/python/inkbox/a2a/resource.py
"""Inkbox serve-side A2A inbox resource."""

from __future__ import annotations

from typing import Any, Iterator

from inkbox._http import HttpTransport
from inkbox.a2a.types import (
    A2AContactRule,
    A2AContext,
    A2AContextPage,
    A2AHistoryDirection,
    A2AHistoryMessage,
    A2AHistoryMessagePage,
    A2AMessageRole,
    A2AReplyIntent,
    A2ARuleAction,
    A2ARuleDirection,
    A2ASettings,
    A2ASkill,
    A2ATask,
    A2ATaskPage,
    A2ATaskState,
    parse_context,
    parse_datetime,
    parse_history_message,
    parse_skill,
    parse_task,
)


class A2AResource:
    """Internal transport wrapper used by :class:`AgentIdentity`."""

    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    @staticmethod
    def _base(handle: str) -> str:
        return f"/identities/{handle}/a2a"

    @staticmethod
    def _rule_direction(
        direction: A2ARuleDirection | str,
    ) -> str:
        value = (
            direction.value
            if isinstance(direction, A2ARuleDirection)
            else direction
        )
        if value not in {"inbound", "outbound", "both"}:
            raise ValueError(
                "A2A rule direction must be 'inbound', 'outbound', or 'both'"
            )
        return value

    def settings(self, handle: str) -> A2ASettings:
        data = self._http.get(f"{self._base(handle)}/settings")
        return A2ASettings(
            enabled=data["enabled"],
            filter_mode=data["filter_mode"],
            skills=(
                [parse_skill(item) for item in data["skills"]]
                if data.get("skills") is not None
                else None
            ),
            card_url=data["card_url"],
            inbound_task_count=data.get("inbound_task_count", 0),
            outbound_task_count=data.get("outbound_task_count", 0),
            updated_at=parse_datetime(data.get("updated_at")),
        )

    def update_settings(self, handle: str, **changes: Any) -> A2ASettings:
        data = self._http.put(f"{self._base(handle)}/settings", json=changes)
        return A2ASettings(
            enabled=data["enabled"],
            filter_mode=data["filter_mode"],
            skills=(
                [parse_skill(item) for item in data["skills"]]
                if data.get("skills") is not None
                else None
            ),
            card_url=data["card_url"],
            inbound_task_count=data.get("inbound_task_count", 0),
            outbound_task_count=data.get("outbound_task_count", 0),
            updated_at=parse_datetime(data.get("updated_at")),
        )

    def card(self, handle: str) -> dict[str, Any]:
        return self._http.get(f"{self._base(handle)}/card")

    def tasks(
        self,
        handle: str,
        *,
        direction: A2AHistoryDirection | str | None = None,
        requester_handle: str | None = None,
        worker_handle: str | None = None,
        state: A2ATaskState | str | None = None,
        context_id: str | None = None,
        q: str | None = None,
        since: str | None = None,
        cursor: str | None = None,
        limit: int = 50,
    ) -> A2ATaskPage:
        data = self._http.get(
            f"{self._base(handle)}/tasks",
            params={
                "direction": (
                    direction.value
                    if isinstance(direction, A2AHistoryDirection)
                    else direction
                ),
                "requester_handle": requester_handle,
                "worker_handle": worker_handle,
                "state": state.value if isinstance(state, A2ATaskState) else state,
                "context_id": context_id,
                "q": q,
                "since": since,
                "cursor": cursor,
                "limit": limit,
            },
        )
        return A2ATaskPage(
            items=[parse_task(item) for item in data["items"]],
            next_cursor=data.get("next_cursor"),
        )

    def iter_tasks(
        self,
        handle: str,
        *,
        direction: A2AHistoryDirection | str | None = None,
        requester_handle: str | None = None,
        worker_handle: str | None = None,
        state: A2ATaskState | str | None = None,
        context_id: str | None = None,
        q: str | None = None,
        since: str | None = None,
        limit: int = 50,
    ) -> Iterator[A2ATask]:
        cursor = None
        while True:
            page = self.tasks(
                handle,
                direction=direction,
                requester_handle=requester_handle,
                worker_handle=worker_handle,
                state=state,
                context_id=context_id,
                q=q,
                since=since,
                cursor=cursor,
                limit=limit,
            )
            yield from page.items
            if not page.next_cursor:
                return
            cursor = page.next_cursor

    def task(self, handle: str, task_id: str) -> A2ATask:
        return parse_task(self._http.get(f"{self._base(handle)}/tasks/{task_id}"))

    def sent_tasks(
        self,
        handle: str,
        *,
        requester_handle: str | None = None,
        worker_handle: str | None = None,
        state: A2ATaskState | str | None = None,
        context_id: str | None = None,
        q: str | None = None,
        since: str | None = None,
        cursor: str | None = None,
        limit: int = 50,
    ) -> A2ATaskPage:
        data = self._http.get(
            f"{self._base(handle)}/sent/tasks",
            params={
                "requester_handle": requester_handle,
                "worker_handle": worker_handle,
                "state": state.value if isinstance(state, A2ATaskState) else state,
                "context_id": context_id,
                "q": q,
                "since": since,
                "cursor": cursor,
                "limit": limit,
            },
        )
        return A2ATaskPage(
            items=[parse_task(item) for item in data["items"]],
            next_cursor=data.get("next_cursor"),
        )

    def iter_sent_tasks(
        self,
        handle: str,
        *,
        requester_handle: str | None = None,
        worker_handle: str | None = None,
        state: A2ATaskState | str | None = None,
        context_id: str | None = None,
        q: str | None = None,
        since: str | None = None,
        limit: int = 50,
    ) -> Iterator[A2ATask]:
        cursor = None
        while True:
            page = self.sent_tasks(
                handle,
                requester_handle=requester_handle,
                worker_handle=worker_handle,
                state=state,
                context_id=context_id,
                q=q,
                since=since,
                cursor=cursor,
                limit=limit,
            )
            yield from page.items
            if not page.next_cursor:
                return
            cursor = page.next_cursor

    def sent_task(self, handle: str, task_id: str) -> A2ATask:
        return parse_task(
            self._http.get(f"{self._base(handle)}/sent/tasks/{task_id}")
        )

    def messages(
        self,
        handle: str,
        *,
        direction: A2AHistoryDirection | str | None = None,
        requester_handle: str | None = None,
        worker_handle: str | None = None,
        task_id: str | None = None,
        context_id: str | None = None,
        role: A2AMessageRole | str | None = None,
        q: str | None = None,
        since: str | None = None,
        cursor: str | None = None,
        limit: int = 50,
    ) -> A2AHistoryMessagePage:
        data = self._http.get(
            f"{self._base(handle)}/messages",
            params={
                "direction": (
                    direction.value
                    if isinstance(direction, A2AHistoryDirection)
                    else direction
                ),
                "requester_handle": requester_handle,
                "worker_handle": worker_handle,
                "task_id": task_id,
                "context_id": context_id,
                "role": role.value if isinstance(role, A2AMessageRole) else role,
                "q": q,
                "since": since,
                "cursor": cursor,
                "limit": limit,
            },
        )
        return A2AHistoryMessagePage(
            items=[parse_history_message(item) for item in data["items"]],
            next_cursor=data.get("next_cursor"),
        )

    def iter_messages(
        self,
        handle: str,
        *,
        direction: A2AHistoryDirection | str | None = None,
        requester_handle: str | None = None,
        worker_handle: str | None = None,
        task_id: str | None = None,
        context_id: str | None = None,
        role: A2AMessageRole | str | None = None,
        q: str | None = None,
        since: str | None = None,
        limit: int = 50,
    ) -> Iterator[A2AHistoryMessage]:
        cursor = None
        while True:
            page = self.messages(
                handle,
                direction=direction,
                requester_handle=requester_handle,
                worker_handle=worker_handle,
                task_id=task_id,
                context_id=context_id,
                role=role,
                q=q,
                since=since,
                cursor=cursor,
                limit=limit,
            )
            yield from page.items
            if not page.next_cursor:
                return
            cursor = page.next_cursor

    def reply(
        self,
        handle: str,
        task_id: str,
        *,
        intent: A2AReplyIntent | str,
        text: str | None = None,
        parts: list[dict[str, Any]] | None = None,
    ) -> A2ATask:
        if (text is None) == (parts is None):
            raise ValueError("Pass exactly one of text or parts")
        reply_parts = [{"text": text}] if text is not None else parts
        data = self._http.post(
            f"{self._base(handle)}/tasks/{task_id}/reply",
            json={
                "intent": intent.value if isinstance(intent, A2AReplyIntent) else intent,
                "parts": reply_parts,
            },
        )
        return parse_task(data)

    def contexts(
        self,
        handle: str,
        *,
        direction: A2AHistoryDirection | str | None = None,
        cursor: str | None = None,
        limit: int = 50,
    ) -> A2AContextPage:
        data = self._http.get(
            f"{self._base(handle)}/contexts",
            params={
                "direction": (
                    direction.value
                    if isinstance(direction, A2AHistoryDirection)
                    else direction
                ),
                "cursor": cursor,
                "limit": limit,
            },
        )
        return A2AContextPage(
            items=[parse_context(item) for item in data["items"]],
            next_cursor=data.get("next_cursor"),
        )

    def context(self, handle: str, context_id: str) -> A2AContext:
        return parse_context(
            self._http.get(f"{self._base(handle)}/contexts/{context_id}")
        )

    def sent_contexts(
        self,
        handle: str,
        *,
        cursor: str | None = None,
        limit: int = 50,
    ) -> A2AContextPage:
        data = self._http.get(
            f"{self._base(handle)}/sent/contexts",
            params={"cursor": cursor, "limit": limit},
        )
        return A2AContextPage(
            items=[parse_context(item) for item in data["items"]],
            next_cursor=data.get("next_cursor"),
        )

    def sent_context(self, handle: str, context_id: str) -> A2AContext:
        return parse_context(
            self._http.get(
                f"{self._base(handle)}/sent/contexts/{context_id}"
            )
        )

    def contact_rules(self, handle: str) -> list[A2AContactRule]:
        return [
            self._parse_rule(item)
            for item in self._http.get(f"{self._base(handle)}/contact-rules")
        ]

    def add_contact_rule(
        self,
        handle: str,
        *,
        peer_handle: str,
        action: A2ARuleAction | str,
        direction: A2ARuleDirection | str = A2ARuleDirection.INBOUND,
    ) -> A2AContactRule:
        data = self._http.post(
            f"{self._base(handle)}/contact-rules",
            json={
                "action": action.value if isinstance(action, A2ARuleAction) else action,
                "match_type": "handle",
                "match_target": peer_handle,
                "direction": self._rule_direction(direction),
            },
        )
        return self._parse_rule(data)

    def update_contact_rule(
        self,
        handle: str,
        rule_id: str,
        *,
        action: A2ARuleAction | str | None = None,
        direction: A2ARuleDirection | str | None = None,
    ) -> A2AContactRule:
        body: dict[str, str] = {}
        if action is not None:
            body["action"] = (
                action.value if isinstance(action, A2ARuleAction) else action
            )
        if direction is not None:
            body["direction"] = self._rule_direction(direction)
        if not body:
            raise ValueError("Pass at least one of action or direction")
        data = self._http.patch(
            f"{self._base(handle)}/contact-rules/{rule_id}",
            json=body,
        )
        return self._parse_rule(data)

    def delete_contact_rule(self, handle: str, rule_id: str) -> None:
        self._http.delete(f"{self._base(handle)}/contact-rules/{rule_id}")

    @staticmethod
    def _parse_rule(data: dict[str, Any]) -> A2AContactRule:
        return A2AContactRule(
            id=data["id"],
            action=A2ARuleAction(data["action"]),
            match_type=data["match_type"],
            match_target=data["match_target"],
            direction=A2ARuleDirection(data["direction"]),
            status=data["status"],
            created_at=parse_datetime(data["created_at"]),  # type: ignore[arg-type]
            updated_at=parse_datetime(data["updated_at"]),  # type: ignore[arg-type]
        )


def skills_wire(skills: list[A2ASkill]) -> list[dict[str, Any]]:
    return [skill.to_wire() for skill in skills]
